File: Assets/Scripts/Block.cs
using UnityEngine;
using System.Collections;

public class Block : MonoBehaviour {
	
	//Block Colorsettings
	public Color blockColor;
	
	//Block Pattern sves here
	public string[] block;
	
	//Flag to trigger if the block should start falling
	public bool playable;
	
	//if its not our regular Blocks then it would be our Nextblock 
	public bool isNextBlock;
	
	
	
	//Private variables
	//this Defines out Block Matrix it would be define as the pattern
	private bool[,] blockMatrix;
	
	//Block Fall speed it would variate as change level or press drop button
	private float fallSpeed;
	
	//Logical position on screen fo block nd its not the Transform position
	private int yPosition;
	private	int xPosition;
	
	//logical Size of The Block 
	private int size;
	
	//this is for defining center origin so block can rotate move 
	private int halfSize;
	private float halfSizeFloat;
	
	
	
	// Use this for initialization
	void Start () {
		
		size = block.Length;
		var width = block[0].Length;
		
		//Code is optional only test certain limitation of block usage		
		if (size < 2) {
			Debug.LogError ("Blocks must have at least two lines");
		return;
		}
		if (width != size) {
			Debug.LogError ("Block width and height must be the same");
		return;
		}
		if (size > GameManager.instance.maxBlockSize) {
			Debug.LogError ("Blocks must not be larger than " + GameManager.instance.maxBlockSize);
		return;
		}
			for ( int i = 1; i < size; i++) {
				if (block[i].Length != block[i-1].Length) {
					Debug.LogError ("All lines in the block must be the same length");
				return;
				}
		}
		////////////////////
		
		
		//To define center origin of the block
		halfSize = ((size*20)/2);
		//Used Float so not to do alot of Type conversions
		halfSizeFloat = ((size*20)/2);
	
		//Block matrix this would allow checking Collistions
		//Basicaly a predefined Physics.
		blockMatrix = new bool[size, size];
		
		//this is where our Pattern is converted into usable Node positions like checking 1 for Node and 0 for empty space
		//This itertes through block pattern array and checks for 1
		for (int y = 0; y < size ; y++) {
			for (int x = 0; x < size; x++) {
				
				if (block[y][x]=="1"[0]) {
					blockMatrix[x, y] = true;
					GameObject blocker =(GameObject) Instantiate(GameManager.instance.Node);
					blocker.transform.parent = transform;
					blocker.transform.localScale=Vector3.one*20;
					//Node placement as our pattern suggests
					blocker.transform.localPosition= new Vector3((x*20)-halfSizeFloat, (size-(y*20))+halfSizeFloat-size, 0.0f);
					//Node each color is now our block color
					blocker.GetComponent<UISprite>().color=this.blockColor;
			
				}
			}
		}
		
		//Check if its our regular Block or Next BLock
		if(!this.isNextBlock)
		{
			//Logical block position from which Block would start falling
			yPosition = (int)((GameManager.instance.fieldHeight*20)-20);
			
			//Pysical Block location with origin
			transform.localPosition =new Vector3((float)((((GameManager.instance.fieldWidth)/2)*20 + (size%2 == 0? 0 : 20/2))),(yPosition - halfSizeFloat),-1);
			
			//Logical X position 
			xPosition =(int)(transform.localPosition.x- halfSizeFloat);
			
			//our fall speed is equal Block Normal speed so current strting speed would Block normal speed
			fallSpeed =(float) GameManager.instance.blockNormalSpeed;
			
			//This is the logic which would define when the block is collided and its th emain source of whole tetris
			if (GameManager.instance.CheckBlock (blockMatrix, (xPosition/20), (yPosition/20))) {
					//Gameover logic initiates when block is unable to move
				GameManager.instance.GameOver();
					
				return;
			}
			
			
			StartCoroutine( Delay(.5f));
			
		}
		
		
	}
	
	
	public IEnumerator Delay (float time) {
	
		yield return new WaitForSeconds(time);
		this.playable=true;
		StartCoroutine(fall());
		StartCoroutine(CheckInput());
	}
	
	public IEnumerator yieldDelay(float time)
	{
		StartCoroutine(Delay(time));
		yield return null;
	}
		


	
	public IEnumerator emptyYield()
	{
		yield return null;
	}
	
public IEnumerator fall () {
	while (playable&&!(GameManager.instance.isGamePaused)) {
		
		yPosition--;//=(int)fallSpeed;
		
		if (GameManager.instance.CheckBlock (blockMatrix, (int)(xPosition/20), (int)(yPosition/20))) {
			GameManager.instance.SetBlock (blockMatrix, (xPosition/20), (yPosition+20)/20, this.blockColor);
		
			Destroy(gameObject);
			
			break;

		}
		
				
		// Make on-screen block fall down 1 square
		// Also serves as a delay...if you want old-fashioned square-by-square movement, replace this with yield WaitForSeconds
		//Our fall logic Forced
		
			
		for (float i = yPosition+1; i > yPosition; i -= Time.deltaTime*(fallSpeed*30)) {
				Vector3 val = transform.localPosition;
				if(transform.localEulerAngles.z==0)
				{
					transform.localPosition=new Vector3(val.x,(i - halfSize),val.z);
				}
				else if(transform.localEulerAngles.z<=180f)
				{
					transform.localPosition=new Vector3(val.x,(i - halfSize)+20,val.z);
					
				}
				else
				{
					transform.localPosition=new Vector3(val.x,(i - halfSize),val.z);
				}
				yield return null;
			
		}
		
	//	StartCoroutine(CheckInput());
		
	}
}
	

	


public IEnumerator CheckInput () {
	
		while(playable&&!(GameManager.instance.isGamePaused))
		{
		
		if (Input.GetKey(KeyCode.LeftArrow)||Input.GetKey(KeyCode.A)) {
			
			MoveHorizontal(-1*20);
			
			yield return new WaitForSeconds(.2f);
			
		}
		else if (Input.GetKey(KeyCode.RightArrow)||Input.GetKey(KeyCode.D)) {
			MoveHorizontal(1*20);
		
			yield return new WaitForSeconds(.2f);
		
		}

		if (Input.GetKey(KeyCode.UpArrow)||Input.GetKey(KeyCode.W)) {
			RotateBlock();
			GameManager.instance.numOfRotations++;
				yield return new WaitForSeconds(.2f);
		}
		
		if (Input.GetKeyDown(KeyCode.DownArrow)||Input.GetKeyDown(KeyCode.S)||Input.GetKeyDown(KeyCode.Space)) {
			GameManager.instance.numOfDropPressed++;
			
		}
		
		if (Input.GetKey(KeyCode.DownArrow)||Input.GetKey(KeyCode.S)||Input.GetKey(KeyCode.Space)) {
			fallSpeed =(float)GameManager.instance.blockDropSpeed;
			GameManager.instance.isDropPressed=true;
			
		}
		else
		{
			fallSpeed =(float) GameManager.instance.blockNormalSpeed ;
			
		}
		
		//Change the next block
		if(Input.GetKeyDown(KeyCode.LeftShift) || Input.GetKeyDown(KeyCode.RightShift))
		{
			GameManager.instance.ChangeNextBlock();
		}
		
		//Get speed down ou up just in Manual game
		if(GameManager.instance.gameKind == "Manually"||GameManager.instance.gameKind=="DynamicM")
		{

			if(Input.GetKeyDown(KeyCode.Plus)||Input.GetKeyDown(KeyCode.KeypadPlus))
			{
				GameManager.instance.blockNormalSpeed += .2;
				GameManager.instance.delayTime -= GameManager.instance.delayTime * 0.4;
				GameManager.instance.numOfSpeedUp++;
			}
			if(((Input.GetKeyDown(KeyCode.Minus)||Input.GetKeyDown(KeyCode.KeypadMinus))) && GameManager.instance.blockNormalSpeed > 1.0)
			{
				GameManager.instance.blockNormalSpeed -= .2;
				GameManager.instance.delayTime += GameManager.instance.delayTime * 0.4;
				GameManager.instance.numOfSpeedDown++;
			}
		}
			yield return null;
		}
		
	
}
	
public void MoveHorizontal (int dir) {
	// Check to see if block could be moved in the desired direction
		
			if (!GameManager.instance.CheckBlock (blockMatrix, (xPosition + dir)/20, (yPosition/20))) {
		
				Vector3 pos = transform.localPosition;
				transform.localPosition=new Vector3((pos.x+dir),pos.y,pos.z);

				xPosition += dir;
				
			}
		
}
	
	

void RotateBlock () {
	
		//This codesmakes the rotation fo logical block
		var tempMatrix = new bool[size, size];
	for (int y = 0; y < size; y++) {
		for (int x = 0; x < size; x++) {
			tempMatrix[y, x] = blockMatrix[x, (size-1)-y];
		}
	}
	
	// If the rotated block doesn't overlap existing blocks, copy the rotated matrix back and rotate on-screen block to match
	if (!GameManager.instance.CheckBlock (tempMatrix, (xPosition/20), (yPosition/20))) {
		System.Array.Copy (tempMatrix, blockMatrix, size*size);
			
				Vector3 pos= this.transform.localPosition;
		
			if(transform.eulerAngles.z==180f)
			{
			
				this.transform.localPosition= new Vector3(pos.x+20,pos.y+20,pos.z);
				
			}
			else if(this.transform.localRotation.z==0f)
			{
			
				this.transform.localPosition= new Vector3(pos.x-20,pos.y,pos.z);
				
			}
			//This line makes the Block rotate Physicaly...
				this.transform.Rotate(Vector3.forward*-90);
			
			
			
			
	}
}
}

File: Assets/Scripts/Block.js
var block : String[];
var playable = true;


private var blockMatrix : boolean[,];
private var fallSpeed : float;
private var yPosition : int;
private var xPosition : int;
private var size : int;
private var halfSize : int;
private var halfSizeFloat : float;
private var dropped = false;
private var material : Material;

function Start () {
	// Sanity checking
	size = block.Length;
	var width = block[0].Length;
	if (size < 2) {
		Debug.LogError ("Blocks must have at least two lines");
		return;
	}
	if (width != size) {
		Debug.LogError ("Block width and height must be the same");
		return;
	}
	if (size > Manager.use.maxBlockSize) {
		Debug.LogError ("Blocks must not be larger than " + Manager.use.maxBlockSize);
		return;
	}
	for (i = 1; i < size; i++) {
		if (block[i].Length != block[i-1].Length) {
			Debug.LogError ("All lines in the block must be the same length");
			return;
		}
	}
	
	halfSize = (size/2);
	halfSizeFloat = (size*.5); // halfSize is an integer for the array, but we need a float for positioning the on-screen cubes (for odd sizes)
	
	// Convert block string array from the inspector to a boolean 2D array for easier usage
	blockMatrix = new boolean[size, size];
	for (y = 0; y < size; y++) {
		for (x = 0; x < size; x++) {
			if (block[y][x] == "1"[0]) {
				blockMatrix[x, y] = true;
				var block = Instantiate(Manager.use.cube, Vector3((x-halfSizeFloat)*.1, ((size-y)+halfSizeFloat-size)*.1, 0.0), Quaternion.identity) as Transform;
				block.parent = transform;
				block.renderer.material = material;
			}
		}
	}
	
	// For blocks with even sizes, we just add 0, but odd sizes need .5 added to the position to work right
	transform.position.x = (Manager.use.FieldWidth()/2 + (size%2 == 0? 0.0 : .5));
	xPosition = transform.position.x - halfSizeFloat;
	yPosition = Manager.use.FieldHeight() - 1;
	transform.position.y = yPosition - halfSizeFloat;
	fallSpeed = Manager.use.blockNormalSpeed;
	
	// Check to see if this block would overlap existing blocks, in which case the game is over
	if (Manager.use.CheckBlock (blockMatrix, xPosition, yPosition)) {
		Manager.use.GameOver();
		return;
	}
	
	CheckInput();
	yield Delay( (1.0 / Manager.use.blockNormalSpeed) * 2.0 );
	Fall();
}

// This is used instead of WaitForSeconds, so that the delay can be cut short if player hits the drop button
function Delay (time : float) {
	var t = 0.0;
	//Debug.Log(time);
	while (t <= Manager.use.delayTime && !dropped) {
		t += Time.deltaTime;	
		yield;
	}
}

function Fall () {
	while (playable) {
		// Check to see if block would collide if moved down one row
		yPosition--;
		if (Manager.use.CheckBlock (blockMatrix, xPosition, yPosition)) {
			Manager.use.SetBlock (blockMatrix, xPosition, yPosition+1, material);
			Destroy(gameObject);
			break;
		}
		
		// Make on-screen block fall down 1 square
		// Also serves as a delay...if you want old-fashioned square-by-square movement, replace this with yield WaitForSeconds
		for (var i : float = yPosition+1; i > yPosition; i -= Time.deltaTime*fallSpeed) {
			transform.position.y = i - halfSizeFloat;
			yield;
		}
	}
}

function CheckInput () {
	while (playable) {
		
		//var input = Input.GetAxis("Horizontal");
		if (Input.GetKey(KeyCode.LeftArrow)||Input.GetKey(KeyCode.A)) {
			yield MoveHorizontal(-1);
		}
		else if (Input.GetKey(KeyCode.RightArrow)||Input.GetKey(KeyCode.W)) {
			yield MoveHorizontal(1);
		}

		if (Input.GetKeyDown(KeyCode.UpArrow)||Input.GetKeyDown(KeyCode.W)) {
			RotateBlock();
		}
		
		if (Input.GetKey(KeyCode.DownArrow)||Input.GetKey(KeyCode.S)) {
			fallSpeed = Manager.use.blockDropSpeed;
			//dropped = true;
			//Manager.use.score += 5;
			//break;	// Break out of while loop, so the coroutine stops (we don't care about input anymore)
		}
		else
		{
			fallSpeed = Manager.use.blockNormalSpeed ;
			
		}
		
		//Change the next block
		if(Input.GetKeyDown(KeyCode.Keypad5) || Input.GetKeyDown(KeyCode.Space))
		{
			Manager.use.ChangeNextBlock();
		}
		
		//Get speed down ou up just in Manual game
		if(Manager.use.gameKind == "Manually")
		{
			if(Input.GetKeyDown(KeyCode.KeypadPlus))
			{
				Manager.use.blockNormalSpeed += .5;
				Manager.use.delayTime -= Manager.use.delayTime * 0.4;
			}
			if(Input.GetKeyDown(KeyCode.KeypadMinus) && Manager.use.blockNormalSpeed > 2.0)
			{
				Manager.use.blockNormalSpeed -= .5;
				Manager.use.delayTime += Manager.use.delayTime * 0.4;
			}
		}
		
		yield;
	}
}

function MoveHorizontal (dir : int) {
	// Check to see if block could be moved in the desired direction
	if (!Manager.use.CheckBlock (blockMatrix, xPosition + dir, yPosition)) {
		transform.position.x += dir;
		xPosition += dir;
		yield WaitForSeconds (Manager.use.blockMoveDelay);
	}
}

function RotateBlock () {
	// Rotate matrix 90° to the right and store the results in a temporary matrix
	var tempMatrix = new boolean[size, size];
	for (y = 0; y < size; y++) {
		for (x = 0; x < size; x++) {
			tempMatrix[y, x] = blockMatrix[x, (size-1)-y];
		}
	}
	
	// If the rotated block doesn't overlap existing blocks, copy the rotated matrix back and rotate on-screen block to match
	if (!Manager.use.CheckBlock (tempMatrix, xPosition, yPosition)) {
		System.Array.Copy (tempMatrix, blockMatrix, size*size);
		transform.Rotate (Vector3.forward * -90.0);
	}
}

function SetMaterial(mat : Material){
	material = mat;
}
